File: backend-rust/download-model/download_model.py
#!/usr/bin/env python3
"""Download the AllMiniLML6V2 ONNX model and set up the HF hub cache structure."""

import hashlib
import os
import sys
import shutil
import urllib.request
import time

BASE_URL = "https://huggingface.co/Qdrant/all-MiniLM-L6-v2-onnx/resolve/main"
FILES = [
    "model.onnx",
    "tokenizer.json",
    "config.json",
    "special_tokens_map.json",
    "tokenizer_config.json",
]


def download_file(url, dest_path, timeout=120):
    """Download a file with retry logic."""
    max_retries = 3
    for attempt in range(max_retries):
        try:
            print(f"  Downloading {os.path.basename(dest_path)}...")
            req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
            with urllib.request.urlopen(req, timeout=timeout) as response:
                total_size = int(response.headers.get("Content-Length", 0))
                downloaded = 0
                chunk_size = 8192 * 64
                with open(dest_path + ".tmp", "wb") as f:
                    while True:
                        chunk = response.read(chunk_size)
                        if not chunk:
                            break
                        f.write(chunk)
                        downloaded += len(chunk)
                        if total_size > 0:
                            percent = int(downloaded * 100 / total_size)
                            if percent % 25 == 0 or downloaded == total_size:
                                print(f"    Progress: {downloaded / 1024 / 1024:.1f}/{total_size / 1024 / 1024:.1f} MB ({percent}%)")
                os.replace(dest_path + ".tmp", dest_path)
                file_size = os.path.getsize(dest_path)
                print(f"    Done ({file_size / 1024 / 1024:.1f} MB)")
                return True
        except Exception as e:
            print(f"    Attempt {attempt + 1}/{max_retries} failed: {e}")
            if os.path.exists(dest_path + ".tmp"):
                os.remove(dest_path + ".tmp")
            if attempt < max_retries - 1:
                time.sleep(2)
    return False


def main():
    if len(sys.argv) < 2:
        print("Usage: python download_model.py <output_cache_dir>")
        sys.exit(1)

    output_dir = sys.argv[1]
    os.makedirs(output_dir, exist_ok=True)

    model_cache_dir = os.path.join(output_dir, "models--Qdrant--all-MiniLM-L6-v2-onnx")
    blobs_dir = os.path.join(model_cache_dir, "blobs")
    refs_dir = os.path.join(model_cache_dir, "refs")
    os.makedirs(blobs_dir, exist_ok=True)
    os.makedirs(refs_dir, exist_ok=True)

    print(f"Downloading model files to {model_cache_dir}...\n")

    # Download all files
    downloaded = {}
    for filename in FILES:
        url = f"{BASE_URL}/{filename}"
        dest_path = os.path.join(blobs_dir, filename)
        print(f"  Target: {url}")
        if download_file(url, dest_path):
            downloaded[filename] = dest_path
        else:
            print(f"  FAILED to download {filename}")
            sys.exit(1)

    # Compute the commit SHA from model.onnx
    model_onnx_path = downloaded["model.onnx"]
    sha256_hash = hashlib.sha256()
    with open(model_onnx_path, "rb") as f:
        while True:
            chunk = f.read(4096)
            if not chunk:
                break
            sha256_hash.update(chunk)
    commit_sha = sha256_hash.hexdigest()
    print(f"\nComputed commit SHA: {commit_sha}")

    # Write refs/main
    refs_main_path = os.path.join(refs_dir, "main")
    with open(refs_main_path, "w") as f:
        f.write(commit_sha)
    print(f"Wrote {refs_main_path}")

    # Create snapshot directory
    snapshot_dir = os.path.join(model_cache_dir, "snapshots", commit_sha)
    os.makedirs(snapshot_dir, exist_ok=True)

    # Copy files to snapshot directory
    for filename, src_path in downloaded.items():
        dest_path = os.path.join(snapshot_dir, filename)
        if not os.path.exists(dest_path):
            shutil.copy2(src_path, dest_path)
            print(f"  Copied {filename} to snapshot")

    print(f"\nAll files downloaded and cached at: {model_cache_dir}")
    print(f"Snapshot directory: {snapshot_dir}")


if __name__ == "__main__":
    main()

File: electron/scripts/download-model.js
const fs = require('fs');
const path = require('path');
const https = require('https');
const { createGunzip } = require('zlib');
const { Extract } = require('tar-stream');
const { pipeline } = require('stream/promises');

const MODEL_CACHE_DIR = path.join(__dirname, '..', '..', 'model_cache');
const HF_HOME = MODEL_CACHE_DIR;
const CACHE_DIR = path.join(HF_HOME, 'hub', 'models--Qdrant--all-MiniLM-L6-v2-onnx');

// Model files to download (from HuggingFace)
const FILES = [
    'model.onnx',
    'config.json',
    'tokenizer.json',
    'tokenizer_config.json',
    'special_tokens_map.json',
];

const BASE_URL = (process.env.HF_ENDPOINT || 'https://huggingface.co') + '/Qdrant/all-MiniLM-L6-v2-onnx/resolve/main';

function downloadFile(url, dest) {
    return new Promise((resolve, reject) => {
        console.log(`  Downloading: ${path.basename(dest)}`);
        const file = fs.createWriteStream(dest);
        https.get(url, { timeout: 60000 }, (response) => {
            if (response.statusCode !== 200) {
                reject(new Error(`HTTP ${response.statusCode} for ${url}`));
                return;
            }
            const total = parseInt(response.headers['content-length'] || '0', 10);
            let downloaded = 0;
            response.on('data', (chunk) => {
                downloaded += chunk.length;
                if (total > 0) {
                    const pct = (downloaded / total * 100).toFixed(1);
                    process.stdout.write(`\r    ${(downloaded/1024/1024).toFixed(1)}MB / ${(total/1024/1024).toFixed(1)}MB (${pct}%)`);
                }
            });
            pipeline(response, file).then(() => {
                process.stdout.write('\n');
                resolve();
            }).catch(reject);
        }).on('error', reject);
    });
}

async function ensureModelDir() {
    const snapshotsDir = path.join(CACHE_DIR, 'snapshots');
    const refsDir = path.join(CACHE_DIR, 'refs');

    // Create directory structure
    fs.mkdirSync(snapshotsDir, { recursive: true });
    fs.mkdirSync(refsDir, { recursive: true });

    // Get the hash from the snapshot directory
    // We'll use a fixed commit hash that corresponds to the model version
    const commitHash = 'bbd7b466f6d58e646fdc2bd5fd67b2f5e93c0b687011bd4548c420f7bd46f0c5';
    const modelDir = path.join(snapshotsDir, commitHash);
    fs.mkdirSync(modelDir, { recursive: true });

    // Write refs/main pointing to the commit hash
    fs.writeFileSync(path.join(refsDir, 'main'), commitHash);
    fs.writeFileSync(path.join(refsDir, 'HEAD'), `ref: refs/heads/main\n`);

    // Download model files
    for (const file of FILES) {
        const filePath = path.join(modelDir, file);
        if (fs.existsSync(filePath)) {
            const stats = fs.statSync(filePath);
            console.log(`  ${file} already exists (${(stats.size/1024/1024).toFixed(1)} MB)`);
            continue;
        }
        const url = `${BASE_URL}/${file}`;
        try {
            await downloadFile(url, filePath);
        } catch (err) {
            console.error(`  Failed to download ${file}: ${err.message}`);
        }
    }

    // Verify
    console.log('\nDownload complete!');
    for (const file of FILES) {
        const filePath = path.join(modelDir, file);
        if (fs.existsSync(filePath)) {
            const stats = fs.statSync(filePath);
            console.log(`  ✓ ${file} (${(stats.size/1024/1024).toFixed(1)} MB)`);
        } else {
            console.log(`  ✗ ${file} not found`);
        }
    }
}

console.log('=== Downloading Embedding Model ===');
console.log(`Model cache dir: ${MODEL_CACHE_DIR}`);
ensureModelDir().catch(err => {
    console.error('Download failed:', err.message);
    process.exit(1);
});